File: 20_valid_parentheses/solution.ts
function isValid(s: string): boolean {
  if (s.length % 2 !== 0) return false;

  const map = {
    '(': ')',
    '[': ']',
    '{': '}',
  };

  let arr: string[] = [];

  for (const cur of s) {
    if (map[cur] !== undefined) {
      arr.push(map[cur]);
    } else {
      // closing brackets
      if (arr.at(-1) !== cur) return false;
      arr.pop();
    }
  }

  return arr.length === 0;
}

File: 20_valid_parentheses/solution.py
from typing import List


class Solution:
    def isValid(self, s: str) -> bool:
        stack: List[str] = []
        parenthesesMap = {")": "(", "]": "[", "}": "{"}

        for c in s:
            if c not in parenthesesMap:
                stack.append(c)
                continue
            if not stack or stack[-1] != parenthesesMap[c]:
                return False
            stack.pop()

        return not stack


solution = Solution()


print(solution.isValid("()"))  # True
print(solution.isValid("([{}])"))  # True
print(solution.isValid("()[]{}"))  # True
print(solution.isValid("(]"))  # False
print(solution.isValid("({)}"))  # False
print(solution.isValid("]"))  # False
